File: prak python/Garage.py
from typing import List
from ParkingLot import ParkingLot
from Car import Car  # Mengimpor kelas Car
from Motorcycle import Motorcycle  # Mengimpor kelas Motorcycle


class Garage(ParkingLot):
    def __init__(self, name: str, area: float, capacity: int):
        super().__init__(capacity)
        self._name = name
        self._area = area
        self._vehicles: List = []  # Mengubah tipe data menjadi List

    def get_name(self):
        return self._name

    def get_area(self):
        return self._area

    def get_vehicles(self):
        return self._vehicles

    def add_vehicle(self, vehicle):
        if isinstance(vehicle, (Car, Motorcycle)):
            self._vehicles.append(vehicle)
            self.num_vehicles += 1  # Menambahkan 1 ke num_vehicles pada ParkingLot
        else:
            raise ValueError(
                "Tidak dapat menambahkan kendaraan yang bukan instance dari Car atau Motorcycle"
            )

    def remove_vehicle(self, vehicle):
        self._vehicles.remove(vehicle)
        self.num_vehicles -= 1  # Mengurangkan 1 dari num_vehicles pada ParkingLot

    def display_vehicles(self):
        for i, vehicle in enumerate(self._vehicles, start=1):
            if isinstance(vehicle, Car):
                print(
                    f"{i}. Car - Brand: {vehicle.brand}, License Plate: {vehicle.license_plate}, Production Year: {vehicle.production_year}, Color: {vehicle.color}, Num Seats: {vehicle.num_seats}, Num Doors: {vehicle.num_doors}"
                )
            elif isinstance(vehicle, Motorcycle):
                print(
                    f"{i}. Motorcycle - Brand: {vehicle.brand}, License Plate: {vehicle.license_plate}, Production Year: {vehicle.production_year}, Color: {vehicle.color}, Motor Type: {vehicle.motor_type}, Tank Capacity: {vehicle.tank_capacity}"
                )
            else:
                print("Unknown Vehicle Type")

File: prak cpp/Garage.cpp
#pragma once

#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include "ParkingLot.cpp"
#include "Vehicle.cpp"
#include "Car.cpp"
#include "Motorcycle.cpp"

using namespace std;

class Garage : public ParkingLot {
private:
    string _name;
    float _area;
    vector<Vehicle*> _vehicles;

public:
    Garage(string name, float area, int capacity)
        : ParkingLot(capacity), _name(name), _area(area) {}

    string get_name() const {
        return _name;
    }

    float get_area() const {
        return _area;
    }

    vector<Vehicle*> get_vehicles() const {
        return _vehicles;
    }

    void add_vehicle(Vehicle* vehicle) {
        if (dynamic_cast<Car*>(vehicle) || dynamic_cast<Motorcycle*>(vehicle)) {
            _vehicles.push_back(vehicle);
            set_num_vehicles(get_num_vehicles() + 1);
        } else {
            throw invalid_argument("Cannot add vehicle that is not an instance of Car or Motorcycle");
        }
    }

    void remove_vehicle(Vehicle* vehicle) {
        auto it = find(_vehicles.begin(), _vehicles.end(), vehicle);
        if (it != _vehicles.end()) {
            _vehicles.erase(it);
            set_num_vehicles(get_num_vehicles() - 1);
        }
    }

    void display_vehicles() const {
        int i = 1;
        for (const auto& vehicle : _vehicles) {
            cout << i << ". ";
            if (dynamic_cast<Car*>(vehicle)) {
                Car* car = dynamic_cast<Car*>(vehicle);
                cout << "Car - Brand: " << car->get_brand() << ", License Plate: " << car->get_license_plate()
                     << ", Production Year: " << car->get_production_year() << ", Color: " << car->get_color()
                     << ", Num Seats: " << car->get_num_seats() << ", Num Doors: " << car->get_num_doors() << endl;
            } else if (dynamic_cast<Motorcycle*>(vehicle)) {
                Motorcycle* motorcycle = dynamic_cast<Motorcycle*>(vehicle);
                cout << "Motorcycle - Brand: " << motorcycle->get_brand() << ", License Plate: " << motorcycle->get_license_plate()
                     << ", Production Year: " << motorcycle->get_production_year() << ", Color: " << motorcycle->get_color()
                     << ", Motor Type: " << motorcycle->get_motor_type() << ", Tank Capacity: " << motorcycle->get_tank_capacity() << endl;
            } else {
                cout << "Unknown Vehicle Type" << endl;
            }
            i++;
        }
    }
};